Source/DafnyCore/AST/Expressions/Expression.cs

```diff
@@ -60,6 +60,14 @@ public abstract class Expression : TokenNode {
       type = value.Normalize();
     }
   }
+  public Type UnnormalizedType {
```

Comment:
Could you add a comment? Also, if there is a way to avoid properties like this I think it would be great, since it's difficult to distinguish variations of a similar thing `UnnormalizedType, Type, Type.Normalize(), Type.NormalizeWrtScope()`

Also, is there a point after which types no longer need to be normalized? I see there are no Normalize calls from the code generators, but `Translator.cs` has one call to `.Normalize()`

Reply:
I would prefer that `.Type` just return whatever value is stored in the `.type` field. However, this seems to have changed at some point, so that `.Type` returns `.type.Normalize()`. I don't quite see the benefit of that, since most places will want to still call `NormalizeExpand()` to expand type synonyms. I assume there was some reason for the change to make `.Type` return the normalized type, so rather than changing that back, I defined the new `UnnormalizedType` field.

I added a comment to `UnnormalizedType`:
The new type inference includes a "type adjustment" phase, which determines the best subset types for a program. This phase works by adjusting (mutating) types in place, using `AdjustableType` type proxies. During that phase, it is necessary to obtain the unnormalized type stored in each AST node, which is what the `UnnormalizedType` property does. This property should only be used during the type adjustment phase. After type inference is complete, use `.Type` instead.

The answer to your question
> Is there a point after which types no longer need to be normalized so we can rely solely on `.Type`?
is that `.Type` should always be used after resolution pass 0. Since `.Type` returns normalized (that is, non-proxy) types, it is then not necessary to call `.Normalize()`. However, in most cases, it will still be necessary to call `.NormalizeExpand()` in order to expand type synonyms.